File: DSA/to_revise/873 - lenLongestFibSubseq.cpp
// https://leetcode.com/problems/length-of-longest-fibonacci-subsequence/

#include <bits/stdc++.h>
using namespace std;

class Solution {
    public:
        int lenLongestFibSubseq(vector<int>& arr) {
            unordered_map<int, int> arr_map;  // Map of elements to their indices
            int res = 0;
            int n = arr.size();
            
            // Create a DP table (2D vector) initialized to 0
            vector<vector<int>> dp(n, vector<int>(n, 0));
    
            // Fill the map with the indices of elements in arr
            for (int i = 0; i < n; ++i) {
                arr_map[arr[i]] = i;
            }
    
            // Iterate in reverse to find Fibonacci subsequences
            for (int i = n - 2; i >= 0; --i) {
                for (int j = n - 1; j > i; --j) {
                    int prev = arr[i], cur = arr[j];
                    int nxt = prev + cur;
                    int length = 2;  // Starting length for Fibonacci sequence (two elements)
    
                    // Check if the next Fibonacci number exists in the array
                    if (arr_map.find(nxt) != arr_map.end()) {
                        int k = arr_map[nxt];  // Get the index of the next number in the sequence
                        length = 1 + dp[j][k];  // Update the length of the subsequence
                        res = max(res, length);  // Track the maximum length
                    }
    
                    dp[i][j] = length;  // Store the length in the DP table
                }
            }
    
            return res;  // Return the result
        }
    };

File: DSA/to_revise/873 - lenLongestFibSubseq.py
# https://leetcode.com/problems/length-of-longest-fibonacci-subsequence/

from typing import List

class Solution:
    def lenLongestFibSubseq(self, arr: List[int]) -> int:
        arr_map = {n: i for i, n in enumerate(arr)}  # map of elements to their indices
        res = 0
        dp = [[0] * len(arr) for _ in range(len(arr))]  # DP table to store lengths

        for i in reversed(range(len(arr) - 1)):
            for j in reversed(range(i + 1, len(arr))):
                prev, cur = arr[i], arr[j]
                nxt = prev + cur
                length = 2  # Fibonacci sequence needs at least two elements

                if nxt in arr_map:  # If the next number exists in the sequence
                    length = 1 + dp[j][arr_map[nxt]]  # Calculate the length
                    res = max(res, length)  # Track the maximum length

                dp[i][j] = length  # Update DP table

        return res
